### PR TITLE
PERF Use a slotmap for hiwire

### PR DESCRIPTION
**Pull request by hoodmane**
A slot map is a map that generates and returns the key when inserting a
value. Slot maps have faster lookup, insertion, and deletion than hash maps,
though they aren't built in to the language so they are a bit more effort to
implement. The implementation here is based roughly on this Rust file:
https://github.com/orlp/slotmap/blob/master/src/basic.rs

This gives an 18% speed improvement on the following:
```js
o = {a: 7}
pyodide.runPython(`
from pyodide.ffi import create_proxy, to_js
from pyodide.code import run_js
from timeit import timeit
n = 1000000
from js import o
def f():
    o.a

timeit("f()", number=n, globals={"f": f}) / n * 1e6
`);
```
and a 14% speed improvement on the following:
```js
pyodide.runPython(`
from pyodide.ffi import create_proxy, to_js
from pyodide.code import run_js
from timeit import timeit
n = 1000000

def f():
    to_js({1: {2: 3}, 4: {5:6, 7:{8:9}}})

timeit("f()", number=n, globals={"f": f}) / n * 1e6
`);
```

This is on top of #4072.

### Checklists

- [x] More comments
- [ ] Add a [CHANGELOG](https://github.com/pyodide/pyodide/blob/main/docs/project/changelog.md) entry

### CODE REVIEW

**Review by ryanking13 — COMMENTED** [review]
Thanks for working on this @hoodmane.

I'm not against this PR, but I think we need more benchmarks before accepting it. I think we need to check things like how much of the total time is taken up by hiwire lookup and insert in common patterns. In addition, slotmap does not seem to be a well-proven data structure.

The code around hiwire is already quite complex, and I'd like to avoid making it even more so unless we have great confidence in the performance gains.

File: src/core/hiwire.c [hunk 1]
```diff
@@ -43,18 +42,114 @@ HIWIRE_INIT_CONSTS();
...
+// If a slot is unoccupied, the multipurpose field of the slotInfo contains the
+// index of the next free slot in the free list or zero if this is the last free
+// slot (for this reason, we do not use slot 0).
+//
+//    unoccupied slot: [version (5 bits)][next free index (25 bits)]0
+//
+// If a slot is occupied, the multipurpose field of the slotInfo contains a 24
+// bit reference count and an IS_DEDUPLICATED bit.
+//
+//      occupied slot: [version (5 bits)][refcount (24 bits)][IS_DEDUPLICATED bit]0
```
[the remark this answers is not in the document]
Comment (hoodmane) [reply src/core/hiwire.c:97]: Yes.

**Comment by hoodmane** [comment]
> I think we need more benchmarks before accepting it

A bunch of profiles are a bit faster, but for instance `to_js` of pyodide-lock.json runs 15 times faster with this PR:
```js
globalThis.lockjs = await (await (fetch(pyodide._api.config.indexURL + "pyodide-lock.json"))).json()
pyodide.runPython(`
from js import lockjs
from pyodide.ffi import to_js
from timeit import timeit

lock = lockjs.to_py()

n = 5000
timeit("to_js(lock)", number=n, globals={"to_js": to_js, "lock": lock}) / n * 1e6
`);
```

**Comment by ryanking13** [comment]
> A bunch of profiles are a bit faster, but for instance to_js of pyodide-lock.json runs 15 times faster with this PR:

Hmmm, that's a very interesting result, thanks for sharing your experiment, I was able to get the same result on multiple local browsers. Maybe hashing keys was a bottleneck?

Since there seems to be a significant performance improvement, I agree with merging this PR, but it would be nice to get a review from @rth as well.

**Comment by hoodmane** [comment]
> Maybe hashing keys was a bottleneck?

Yeah it seems that accessing/modifying a map is very very slow compared to accessing/modifying an array. which in principle isn't surprising since arrays are much simpler data structures. what's pretty fascinating is that I had:
```js
delete _hiwire.objects[index]
```
and this was super slow so I replaced it with:
```js
_hiwire.objects[index] = undefined;
```
which is 100x faster...

**Review by rth — COMMENTED** [review]
Thanks for doing this work! I agree with the fact that we need a better alternative to a JS map, and benchmarks show that it's a real bottleneck. But also I concur with @ryanking13 that hiwire is already complex enough and this adds a whole new level of complexity with a low-level implementation of the slotmap. So if this code needs to be modified by anyone else than Hood in the future that would certainly be an issue.

Just taking a step back, there is no way we could vendor an existing implementation for a faster map and use that? For instance, there is https://github.com/SergeyMakeev/slot_map though it's C++. Otherwise, maybe even  https://github.com/troydhanson/uthash would be comparable in practice unless there is really a need to use slotmap?

File: src/core/hiwire.c [hunk 2]
```diff
@@ -32,29 +32,121 @@ HIWIRE_INIT_CONSTS();
...
+//    unoccupied slot: [version (5 bits)][next free index (25 bits)]0
+//
+// If a slot is occupied, the multipurpose field of the slotInfo contains a 24
+// bit reference count and an IS_DEDUPLICATED bit.
+//
+//      occupied slot: [version (5 bits)][refcount (24 bits)][IS_DEDUPLICATED bit]1
+//
+// References used by JsProxies are deduplicated which makes allocating/freeing
+// them more expensive.
+
```
[the remark this answers is not in the document]
Comment (hoodmane) [reply src/core/hiwire.c:101]: Oh yes, good idea to add this link in the comments, I put it in the PR description:
[orlp/slotmap@master/src/basic.rs](https://github.com/orlp/slotmap/blob/master/src/basic.rs?rgh-link-date=2023-08-18T14%3A47%3A41Z)

**Comment by rth** [comment]
OK, thanks for the explanations: If we go with a re-implementation of slotmap, could we at least abstract it a bit more? So it's defined in a separate file and the implementation details and those macros don't leak so much into hiwire. Currently, with the added macros, it's hard to know what's really needed by hiwire and what's an implementation internals of slotmap. Or do you think it would hurt performance?

**Comment by hoodmane** [comment]
> those macros don't leak so much into hiwire

It is probably a good idea to think more about how to divide the abstraction layers. There are a few layers one can think about here:

1. the slotmap itself
2. the core functions `new_value`, `incref`, `decref`, `get_value`, `incref_deduplicate`
3. stack and immortal references
4. the rest of hiwire which is a collection of utility functions for doing JavaScript stuff from C

Layers 1 and 2 benefit a lot from very deep integration, so much so that they aren't distinct at all. For example, ordinary slotmaps usually don't have reference counting built in, but I think making the slotmap come with refcounts improves the efficiency a fair amount in our case. Integrating the deduplication as well also gives some extra performance wins.

Layer 3 needs the references produced by layers 1+2 to always be odd to stay out of the way, so there is some design interaction, but it is also pretty mild.

Layer 4 is totally independent, and perhaps should be separate from the "hiwire" namespace.

**Comment by hoodmane** [comment]
I found out that externref support has been merged to llvm as of June 10th and we can use it since Emscripten 3.1.42. I talked to @antocuni and my plan is now to unvendor this slotmap logic into a small separate library called hiwire which just has this slotmap logic. I made a repo for it here: https://github.com/hoodmane/hiwire

**Review by ryanking13 — APPROVED** [review]
> So can we merge this for the 0.24 release and plan to unvendor it after the release?

That's fine with me. Thanks for your work again!

**Comment by rth** [comment]
Thanks, sounds good as well.

**Pull request merged**